### PR TITLE
Improved camera controls

### PR DESCRIPTION
**Pull request by reuben-thomas**
### Implemented feature

Improving camera controls for the cursor to address https://github.com/open-rmf/rmf_site/issues/75 based on picking

### Implementation description

I'm going to refactor the camera_controls module to distinguish commands from the mouse, keyboard (https://github.com/open-rmf/rmf_site/issues/98) and eventually touchscreen (https://github.com/open-rmf/rmf_site/issues/97). This should allow us to separately configure each input type and manage conflicts.

### CODE REVIEW

File: rmf_site_editor/src/interaction/camera_controls/cursor.rs [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+            let denom = Vec3::Z.dot(cursor_ray.direction());
+            if denom.abs() > f32::EPSILON && pitch.abs() >= MIN_SELECTION_ANGLE {
+                let dist = (-1.0 * cursor_ray.origin()).dot(Vec3::Z) / denom;
+                if dist > f32::EPSILON {
+                    return cursor_ray.origin() + cursor_ray.direction() * dist;
+                }
+            }
+
+            // No groundplane intersection
+            // Pick a point of a virtual sphere around the camera, of same radius as its height
```
Comment (mxgrey) [inline rmf_site_editor/src/interaction/camera_controls/cursor.rs:345]: Using the height of the camera as the radius of the maximum sphere is an interesting choice. Is there a specific rationale behind that, or is it just based on experimenting with different ideas and seeing what feels best?

Comment (reuben-thomas) [reply rmf_site_editor/src/interaction/camera_controls/cursor.rs:345]: My intuition was that when the cursor is grabbing nothing, it should behave as if selecting the nearest object, which would be the ground. In hindsight, I think a better way to approximate the nearest object would be to grab the nearest [visible entity](https://docs.rs/bevy/latest/bevy/core_pipeline/core_3d/struct.Camera3dBundle.html#structfield.visible_entities) from the active camera

File: rmf_site_editor/src/interaction/camera_controls/cursor.rs [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+    mouse_input: Res<Input<MouseButton>>,
+    keyboard_input: Res<Input<KeyCode>>,
+    raycast_sources: Query<&RaycastSource<SiteRaycastSet>>,
+    cameras: Query<(&Projection, &Transform, &GlobalTransform)>,
+    primary_windows: Query<&Window, With<PrimaryWindow>>,
+) {
+    if let Ok(window) = primary_windows.get_single() {
+        // Return if cursor not within window
+        if window.cursor_position().is_none() {
+            *cursor_command = CursorCommand::default();
```
Comment (mxgrey) [inline rmf_site_editor/src/interaction/camera_controls/cursor.rs:72]: This line causes a jarring drop out of the control mode if the cursor ever leaves the screen space.

I've opened a PR targeting your fork with a fix that I would recommend: https://github.com/reuben-thomas/rmf_site/pull/1

File: rmf_site_editor/src/interaction/camera_controls/keyboard.rs [hunk 3]
```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
```
Comment (mxgrey) [inline rmf_site_editor/src/interaction/camera_controls/keyboard.rs:2]: Nitpick: 2024

File: rmf_site_editor/src/interaction/camera_controls/keyboard.rs [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+        let mut target_zoom_motion = 0.0;
+        if keyboard_input.pressed(KeyCode::N) {
+            target_zoom_motion += -1.0;
+        }
+        if keyboard_input.pressed(KeyCode::M) {
+            target_zoom_motion += 1.0;
+        }
+        // Temporarily use M, N as zoom keys since +-, <>, [] cannot be detected with shift
+        // keyboard_input.get_pressed().for_each(|key| {
```
Comment (mxgrey) [inline rmf_site_editor/src/interaction/camera_controls/keyboard.rs:102]: Looks like this is dead commented code, so it can be removed?

**Review by mxgrey — CHANGES_REQUESTED** [review]
This is really awesome! The camera interaction is much smoother and more intuitive with these improvements.

I'm just leaving a few small comments, in addition to the things we've already discussed offline.

File: rmf_site_editor/src/interaction/camera_controls/keyboard.rs [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        }
+        if keyboard_input.pressed(KeyCode::Right) {
+            target_keyboard_motion.x += 1.0;
+        }
+        if target_keyboard_motion.length() > 0.0 {
+            target_keyboard_motion = target_keyboard_motion.normalize();
+        }
+
+        let mut target_zoom_motion = 0.0;
+        if keyboard_input.pressed(KeyCode::N) {
```
Comment (mxgrey) [inline rmf_site_editor/src/interaction/camera_controls/keyboard.rs:113]: I think `M` and `N` are likely to be used for other hotkeys. I'd suggest changing this to `PgUp` and `PgDn`.

File: rmf_site_editor/src/interaction/camera_controls/mod.rs [hunk 6]
```diff
@@ -481,73 +489,70 @@ fn camera_controls(
...
-
-            if changed {
-                // emulating parent/child to make the yaw/y-axis rotation behave like a turntable
-                // parent = x and y rotation
-                // child = z-offset
-                let rot_matrix = Mat3::from_quat(persp_transform.rotation);
-                persp_transform.translation = controls.orbit_center
-                    + rot_matrix.mul_vec3(Vec3::new(0.0, 0.0, controls.orbit_radius));
+        // Panning
+        } else if (cursor_command.command_type == CameraCommandType::Pan) {
```
Comment (mxgrey) [inline rmf_site_editor/src/interaction/camera_controls/mod.rs:547]: The compiler is flagging this with a warning about unnecessary parentheses.

File: rmf_site_editor/src/interaction/camera_controls/keyboard.rs (hunk 5 above)
Comment (arjo129) [reply rmf_site_editor/src/interaction/camera_controls/keyboard.rs:113]: Knit: Smaller keyboards may not have `PgUp` and `PgDn`.

Comment (mxgrey) [reply rmf_site_editor/src/interaction/camera_controls/keyboard.rs:113]: I think there's usually an equivalent shortcut, often using the `Fn` key.

In the long-term we should really have a hotkey register that lets people customize what keys are used for hotkeys, but in the meantime I think assigning the zoom in/out feature to keys that aren't always available is okay. If someone doesn't have `PgUp`/`PgDn` they will almost certainly still have some way to scroll, and they'll be able to use that for zooming.

**Review by mxgrey — APPROVED** [review]
Thanks for these huge UX improvements!

**Pull request merged**